shadowsocks/stream.go

```diff
@@ -318,31 +331,38 @@ func (cr *chunkReader) readMessage(buf []byte) error {
 	return nil
 }
 
+// ReadChunk returns the next chunk from the stream.  Callers must fully
+// consume and discard the previous chunk before calling ReadChunk again.
 func (cr *chunkReader) ReadChunk() ([]byte, error) {
 	if err := cr.init(); err != nil {
 		return nil, err
 	}
+
+	// Release the previous payload buffer.
+	cr.payload.Release()
+
 	// In Shadowsocks-AEAD, each chunk consists of two
 	// encrypted messages.  The first message contains the payload length,
-	// and the second message is the payload.
-	sizeBuf := cr.buf[:2+cr.aead.Overhead()]
-	if err := cr.readMessage(sizeBuf); err != nil {
+	// and the second message is the payload.  Idle read threads will
+	// block here until the next chunk.
+	if err := cr.readMessage(cr.size); err != nil {
```

Comment:
Perhaps call Acquire here, so you don't need two buffers?

Reply:
The comment here is trying to explain why that doesn't work.  This is the point at which idle read threads block, so we don't want to be holding any more memory than necessary.